metadata/src/main/java/org/apache/kafka/timeline/TimelineHashMap.java

```diff
@@ -0,0 +1,414 @@
+            this.startEpoch = startEpoch;
+        }
+
+        @Override
+        public long startEpoch() {
+            return startEpoch;
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof TimelineHashMapEntry)) return false;
+            TimelineHashMapEntry<K, V> other = (TimelineHashMapEntry<K, V>) o;
+            return key.equals(other.key);
+        }
+
+        @Override
+        public int hashCode() {
+            return key.hashCode();
+        }
+    }
+
+    public TimelineHashMap(SnapshotRegistry snapshotRegistry, int expectedSize) {
+        super(snapshotRegistry, expectedSize);
+    }
+
+    @Override
+    public int size() {
+        return size(SnapshottableHashTable.LATEST_EPOCH);
+    }
+
+    public int size(long epoch) {
+        return snapshottableSize(epoch);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return isEmpty(SnapshottableHashTable.LATEST_EPOCH);
+    }
+
+    public boolean isEmpty(long epoch) {
+        return snapshottableSize(epoch) == 0;
+    }
+
+    @Override
+    public boolean containsKey(Object key) {
+        return containsKey(key, SnapshottableHashTable.LATEST_EPOCH);
+    }
+
+    public boolean containsKey(Object key, long epoch) {
+        return snapshottableGet(new TimelineHashMapEntry<>(key, null), epoch) != null;
+    }
+
+    @Override
+    public boolean containsValue(Object value) {
+        Iterator<Entry<K, V>> iter = entrySet().iterator();
+        while (iter.hasNext()) {
+            Entry<K, V> e = iter.next();
+            if (value.equals(e.getValue())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public V get(Object key) {
+        return get(key, SnapshottableHashTable.LATEST_EPOCH);
+    }
+
+    public V get(Object key, long epoch) {
+        Entry<K, V> entry =
+            snapshottableGet(new TimelineHashMapEntry<>(key, null), epoch);
+        if (entry == null) {
+            return null;
+        }
+        return entry.getValue();
+    }
+
+    @Override
+    public V put(K key, V value) {
+        Objects.requireNonNull(key);
+        Objects.requireNonNull(value);
+        TimelineHashMapEntry<K, V> entry = new TimelineHashMapEntry<>(key, value);
+        TimelineHashMapEntry<K, V> prev = snapshottableAddOrReplace(entry);
+        if (prev == null) {
+            return null;
+        }
+        return prev.getValue();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public V remove(Object key) {
+        return (V) snapshottableRemove(new TimelineHashMapEntry<>(key, null));
+    }
+
+    @Override
+    public void putAll(Map<? extends K, ? extends V> map) {
+        for (Map.Entry<? extends K, ? extends V> e : map.entrySet()) {
+            put(e.getKey(), e.getValue());
+        }
+    }
+
+    @Override
+    public void clear() {
```

Comment:
> clear() just removes the current elements in the map, but doesn't change or delete snapshots.

Hmm, could we document that? Also, do we have a use case for that? Once we clear the top tier, the snapshot may no longer be complete. So, not sure how useful it is.

Reply:
Nothing we do to the current state changes the snapshot state.  Only deleting a snapshot can change snapshot state (by expunging it completely)  Snapshots are not mutable.